### DISCUSSION TITLE
Can I get inverse after resample?

### DISCUSSION THREAD

**Discussion by Bigsealion** [TorchIO-project/torchio#856] [Q&A] [answered]
Hi
I have read this [discussion about resample](https://github.com/fepegar/torchio/discussions/493), and I want to know, is there a method to do something like resample (For example, affine transformations) but invertible?

That's because, in my experiment, the low resolution but complete input is superior to the raw resolution but the patch-based approach. But, I need to inverse the low-resolution output to native space, but I can't easily do this by `torchio`.

In addition, applying `resample` (input native space information) will get a low-resolution output like a  Mosaic...... So I think that's not an ideal solution.

Thanks!

**Comment by fepegar** [comment]
Hi, @Bigsealion.

> I have read this https://github.com/fepegar/torchio/discussions/493, and I want to know, is there a method to do something like resample (For example, affine transformations) but invertible?

The affine transform is invertible.

> That's because, in my experiment, the low resolution but complete input is superior to the raw resolution but the patch-based approach. But, I need to inverse the low-resolution output to native space, but I can't easily do this by torchio.

I'm not sure I understand what you mean by "complete input", "raw resolution". Can you maybe post some screenshots?

**Comment by Bigsealion** [comment]
Hi @fepegar 

I'm so sorry for my poor expression.

> I'm not sure I understand what you mean by "complete input", "raw resolution". Can you maybe post some screenshots?

I have a 3d brain image, but I can't input it to CNN with true spacing ("raw resolution"), so I need to use resample to get a low-resolution 3d image, which can be input to a 3d CNN. And the "low-resolution image" is what I called "complete input", which is the opposite of "patch-based input".

And this is a screenshot for "complete input" and "raw resolution"

"complete input" (means I can take the full image as input): 
![image](https://user-images.githubusercontent.com/23148550/162598243-2e49243c-d0dc-48b6-8d3f-857108be5d0c.png)

"raw resolution" (I can not input the full image to a CNN):
![image](https://user-images.githubusercontent.com/23148550/162598191-41084e34-eabb-43e0-86c6-97466d8d54a4.png)

> The affine transform is invertible.

Yes, but I want to know can the `affine` achieve a similar effect to the `resample`?

Thanks!

**Reply by Bigsealion** [reply]
The main problem I'm having right now is that I cannot inverse the resampled image to the native space.

I look forward to your reply.

Thanks!

**Comment by romainVala** [comment, marked as answer]
Hi
not sure to fully understand, but if you want to get back to the original resolution, you can use what fernando proposed here
https://github.com/fepegar/torchio/discussions/493#discussioncomment-543940
`tio.Resample(original_image)(prediction)`

**Reply by Bigsealion** [reply]
Thanks for your reply!

This makes the output look like a Mosaic. But there may be no better way...

Thanks!

**Comment by ghost** [comment]
Hi all.
Please can one get the inverse of tio.ZNormalization(masking_method=tio.ZNormalization.mean) and tio.HistogramStandardization(landmarks)

**Reply by romainVala** [reply]
Hello
no those two transform are curently not invertible. Not sure about the histrogramStandardization but for ZNormalization it should be easy to implement (on need to add the mean and the std in the transform's history (ie adding to `self.args_names `)
it you want to propose a PR ...
